# Scrape tweets and save as txt files

In [0]:
pip install getoldtweets3

In [0]:
import GetOldTweets3 as got
def scrape_tweets(username):
  username = username
  count = 2000
  # Creation of query object
  tweetCriteria = got.manager.TweetCriteria().setUsername(username)\
                                          .setMaxTweets(count)
  # Creation of list that contains all tweets
  tweets = got.manager.TweetManager.getTweets(tweetCriteria)
  # Creating list of chosen tweet data
  user_tweets = [[tweet.text] for tweet in tweets]
  return user_tweets

In [0]:
# Return tweets of Trump and Obama

#realDonaldTrump
#BarackObama

trump = scrape_tweets('@realDonaldTrump')
obama = scrape_tweets('@BarackObama')

obama

ERROR! Session/line number was not unique in database. History logging moved to new session 60


[["It's Census Day and we all have a crucial role to play. Filling out the 2020 Census determines what the next decade will look like for your community: your roads, health care, schools, representation and more. Takes a few minutes by mail, phone or online. http://my2020census.gov/"],
 ["We've seen all too terribly the consequences of those who denied warnings of a pandemic. We can't afford any more consequences of climate denial. All of us, especially young people, have to demand better of our government at every level and vote this fall."],
 ['So many Americans right now are stepping up to help their communities. Zohaib Begg, a 7-year-old from VA, is one of them. This inspiring kid has been gathering supplies from hotels to donate to the medical professionals at his local hospital.'],
 ['We are grateful for what he did for this country. He carried the baton longer and surer than almost anybody. It falls to the rest of us now to pick it up and never stop moving forward until we finis

In [0]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
# Save all tweets to two folders
for i in range(0,len(trump)):
  path='/content/drive/Tweets Trump vs. Obama/Trump/trump'+str(i)+'.txt'
  with open(path, "w") as output:
    output.write(str(trump[i]))
    print(i)
    next


0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
27

In [0]:
for i in range(0,len(obama)):
  path2='/content/drive/Tweets Trump vs. Obama/Obama/obama'+str(i)+'.txt'
  with open(path2, "w") as output:
    output.write(str(obama[i]))
    print(i)
    next

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
27

# Loading data and prepare data for modeling

In [0]:
#Load the data saved from drive


import os

imdb_dir = '/content/drive/Tweets Trump vs. Obama/'

labels = []
texts = []
count = 0
for label_type in ['Trump/', 'Obama/']:
    dir_name = os.path.join(imdb_dir, label_type)
    for fname in os.listdir(dir_name):
        if fname[-4:] == '.txt':
            count = count + 1
            print(count)
            f = open(os.path.join(dir_name, fname))
            texts.append(f.read())
            f.close()
            if label_type == 'Trump/':
                labels.append(0)
            else:
                labels.append(1)

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
277


In [0]:
# Tokenize, pad and prepare training and validation data
from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences
import numpy as np

maxlen = 100  # cut tweets after 100 words
training_samples = 2560  # training set includes 200 samples
validation_samples = 640  # validation set includes 10000 samples
max_words = 10000  #only consider the top 10,000 words in the dataset  


# Vectorize a text corpus by turning each text into either a sequence of integers
# omits common characters 
tokenizer = Tokenizer(num_words=max_words)

# fit_on_texts to learn the indices of the words
tokenizer.fit_on_texts(texts)

# the sequences are comprised of those indices
sequences = tokenizer.texts_to_sequences(texts)

word_index = tokenizer.word_index
print('Found %s unique tokens.' % len(word_index))

# pad sequences to gain uniform length in case of low word counts
data = pad_sequences(sequences, maxlen=maxlen)

# converted labels from list to a numpy array for modeling
labels = np.asarray(labels)
print('Shape of data tensor:', data.shape)
print('Shape of label tensor:', labels.shape)

# ensure that all sequences in a list have the same length


Found 10172 unique tokens.
Shape of data tensor: (4000, 100)
Shape of label tensor: (4000,)


In [0]:
# Shuffle the data, split the data into a training set and a validation set
indices = np.arange(data.shape[0])
np.random.shuffle(indices)
data = data[indices]
labels = labels[indices]

x_train = data[:training_samples] # from 0 to (2560) training samples
y_train = labels[:training_samples]
x_val = data[training_samples: training_samples + validation_samples]
y_val = labels[training_samples: training_samples + validation_samples]

In [0]:
# save list into file for easily read

np.ndarray.tofile(data,"data1.npy")
np.ndarray.tofile(labels,"labels1.npy")

In [0]:
x_test = data[training_samples + validation_samples:]
y_test = labels[training_samples + validation_samples:]

# Modeling: Fit three models with different architecture

In [0]:
# Model1. LSTM + GRU

from keras.models import Sequential
from keras.layers import Embedding, Flatten, Dense
import tensorflow as tf
from tensorflow.keras.callbacks import EarlyStopping

embedding_dim = 8
#input_shape=(max_words, maxlen*embedding_dim),
model1 = tf.keras.Sequential()
model1.add(tf.keras.layers.Embedding(max_words, embedding_dim, input_length=maxlen))
model1.add(tf.keras.layers.LSTM(80, activation='relu',return_sequences=True))
model1.add(tf.keras.layers.GRU(40))
model1.add(tf.keras.layers.Flatten())
model1.add(tf.keras.layers.Dense(32, activation='relu'))
model1.add(tf.keras.layers.Dense(1, activation='sigmoid'))
model1.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['acc'])
model1.summary()


es = EarlyStopping(monitor='val_loss', mode='min',
                   patience=10, 
                   verbose=1)

history = model1.fit(x_train, y_train,
                    epochs=30,
                    batch_size=20,
                    validation_data=(x_val, y_val),
                    verbose=1,
                    callbacks=[es])


Model: "sequential_31"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_29 (Embedding)     (None, 100, 8)            80000     
_________________________________________________________________
lstm_12 (LSTM)               (None, 100, 80)           28480     
_________________________________________________________________
gru_3 (GRU)                  (None, 40)                14640     
_________________________________________________________________
flatten_5 (Flatten)          (None, 40)                0         
_________________________________________________________________
dense_11 (Dense)             (None, 32)                1312      
_________________________________________________________________
dense_12 (Dense)             (None, 1)                 33        
Total params: 124,465
Trainable params: 124,465
Non-trainable params: 0
_______________________________________________

Mean value is:  0.47372496 ; Median value is:  0.01744634 Predict size is: (800, 1)


In [0]:
from sklearn.metrics import classification_report

y_pred_1 = model1.predict(x_test)

predict_1 = []
for i in y_pred_1:
  if i >= np.mean(y_pred_1):
    predict_1.append(1)
  else:
    predict_1.append(0)

print ("Mean value is: ", np.mean(y_pred_1),";", "Median value is: ", np.median(y_pred_1), "Predict size is:", y_pred_1.shape)
print ("Classify by Mean value", np.mean(y_pred_1),":")
print(classification_report(y_test, predict_1))


predict_1_2 = []
for i in y_pred_1:
  if i >= np.median(y_pred_1):
    predict_1_2.append(1)
  else:
    predict_1_2.append(0)
print ("Classify by Median value", np.median(y_pred_1),":")
print(classification_report(y_test, predict_1_2))

Mean value is:  0.46589142 ; Median value is:  0.4658914 Predict size is: (800, 1)
Classify by Mean value 0.46589142 :
              precision    recall  f1-score   support

           0       0.52      1.00      0.68       412
           1       0.00      0.00      0.00       388

    accuracy                           0.52       800
   macro avg       0.26      0.50      0.34       800
weighted avg       0.27      0.52      0.35       800

Classify by Median value 0.4658914 :
              precision    recall  f1-score   support

           0       0.50      0.06      0.11       412
           1       0.48      0.94      0.64       388

    accuracy                           0.48       800
   macro avg       0.49      0.50      0.37       800
weighted avg       0.49      0.48      0.37       800



/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [0]:

# Model2. LSTM 

embedding_dim = 8
model2 = tf.keras.Sequential()
model2.add(tf.keras.layers.Embedding(max_words,embedding_dim, input_length=maxlen))
model2.add(tf.keras.layers.LSTM(80, activation='relu',return_sequences=True))
model2.add(tf.keras.layers.Flatten())
model2.add(tf.keras.layers.Dense(32, activation='relu'))
model2.add(tf.keras.layers.Dense(1, activation='sigmoid'))
model2.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['acc'])
model2.summary()


es = EarlyStopping(monitor='val_loss', mode='min',
                   patience=10, 
                   verbose=1)

history2 = model2.fit(x_train, y_train,
                    epochs=30,
                    batch_size=20,
                    validation_data=(x_val, y_val),
                    verbose=1,
                    callbacks=[es])


Model: "sequential_23"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_21 (Embedding)     (None, 100, 8)            80000     
_________________________________________________________________
lstm_8 (LSTM)                (None, 100, 80)           28480     
_________________________________________________________________
flatten_3 (Flatten)          (None, 8000)              0         
_________________________________________________________________
dense_6 (Dense)              (None, 32)                256032    
_________________________________________________________________
dense_7 (Dense)              (None, 1)                 33        
Total params: 364,545
Trainable params: 364,545
Non-trainable params: 0
_________________________________________________________________
Epoch 1/30
128/128 [==============================] - 7s 56ms/step - loss: 0.4859 - acc: 0.7773 - val_loss: 0.321

In [0]:
y_pred_2 = model2.predict(x_test)

predict_2 = []
for i in y_pred_2:
  if i >= np.mean(y_pred_2):
    predict_2.append(1)
  else:
    predict_2.append(0)

print ("Mean value is: ", np.mean(y_pred_2),";", "Median value is: ", np.median(y_pred_2), "Predict size is:", y_pred_2.shape)
print ("Classify by Mean value", np.mean(y_pred_2),":")
print(classification_report(y_test, predict_2))


predict_2_2 = []
for i in y_pred_2:
  if i >= np.median(y_pred_2):
    predict_2_2.append(1)
  else:
    predict_2_2.append(0)
print ("Classify by Median value", np.median(y_pred_2),":")
print(classification_report(y_test, predict_2_2))

Mean value is:  0.47372496 ; Median value is:  0.01744634 Predict size is: (800, 1)
Classify by Mean value 0.47372496 :
              precision    recall  f1-score   support

           0       0.94      0.96      0.95       412
           1       0.96      0.94      0.95       388

    accuracy                           0.95       800
   macro avg       0.95      0.95      0.95       800
weighted avg       0.95      0.95      0.95       800

Classify by Median value 0.01744634 :
              precision    recall  f1-score   support

           0       0.95      0.93      0.94       412
           1       0.93      0.95      0.94       388

    accuracy                           0.94       800
   macro avg       0.94      0.94      0.94       800
weighted avg       0.94      0.94      0.94       800



In [0]:
# Model3. Conv1D + LSTM

model3 = tf.keras.Sequential()

model3.add(tf.keras.layers.Embedding(max_words, embedding_dim, input_length=maxlen))

model3.add(tf.keras.layers.Conv1D(filters=80, kernel_size=3,activation='relu')) 
model3.add(tf.keras.layers.MaxPooling1D(pool_size=3))
model3.add(tf.keras.layers.Dropout(0.5))

model3.add(tf.keras.layers.LSTM(50, activation='relu'))
model3.add(tf.keras.layers.Dropout(0.5))

model3.add(tf.keras.layers.Flatten())
model3.add(tf.keras.layers.Dense(64, activation='relu'))
model3.add(tf.keras.layers.Dense(32, activation='relu'))
model3.add(tf.keras.layers.Dropout(0.5))
model3.add(tf.keras.layers.Dense(1, activation='sigmoid'))
model3.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['acc'])
model3.summary()


es = EarlyStopping(monitor='val_loss', mode='min',
                   patience=10, 
                   verbose=1)

history3 = model3.fit(x_train, y_train,
                    epochs=30,
                    batch_size=20,
                    validation_data=(x_val, y_val),
                    verbose=1,
                    callbacks=[es])



Model: "sequential_30"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_28 (Embedding)     (None, 100, 8)            80000     
_________________________________________________________________
conv1d_4 (Conv1D)            (None, 98, 80)            2000      
_________________________________________________________________
max_pooling1d_2 (MaxPooling1 (None, 32, 80)            0         
_________________________________________________________________
dropout_4 (Dropout)          (None, 32, 80)            0         
_________________________________________________________________
lstm_11 (LSTM)               (None, 50)                26200     
_________________________________________________________________
dropout_5 (Dropout)          (None, 50)                0         
_________________________________________________________________
flatten_4 (Flatten)          (None, 50)              

In [0]:
y_pred_3 = model3.predict(x_test)

predict_3 = []
for i in y_pred_3:
  if i >= np.mean(y_pred_3):
    predict_3.append(1)
  else:
    predict_3.append(0)

print ("Mean value is: ", np.mean(y_pred_3),";", "Median value is: ", np.median(y_pred_3), "Predict size is:", y_pred_3.shape)
print ("Classify by Mean value", np.mean(y_pred_3),":")
print(classification_report(y_test, predict_3))


predict_3_2 = []
for i in y_pred_3:
  if i >= np.median(y_pred_3):
    predict_3_2.append(1)
  else:
    predict_3_2.append(0)
print ("Classify by Median value", np.median(y_pred_3),":")
print(classification_report(y_test, predict_3_2))


Mean value is:  0.48680103 ; Median value is:  0.03797695 Predict size is: (800, 1)
Classify by Mean value 0.48680103 :
              precision    recall  f1-score   support

           0       0.95      0.94      0.95       412
           1       0.94      0.94      0.94       388

    accuracy                           0.94       800
   macro avg       0.94      0.94      0.94       800
weighted avg       0.94      0.94      0.94       800

Classify by Median value 0.03797695 :
              precision    recall  f1-score   support

           0       0.96      0.86      0.91       412
           1       0.87      0.96      0.91       388

    accuracy                           0.91       800
   macro avg       0.91      0.91      0.91       800
weighted avg       0.91      0.91      0.91       800



# Analysis: Compare the results

1. The 'support' index shows the number of occurrence of 2 classes are: 412 and 388, which is a comparatively balanced data set.

2. According to precision score, 1st modeel correcty clarified 52% of Trump's tweets if setting cut-off value based on mean value of prediction values. While correctly predict null of Obama's tweets. So I additionally set median value as cut-off. The result shows more about this model: either for Trump or Obama, 1st model could only predict less than 50% of their tweets.

3. Then I tried to simplify my model by taking off the GRU layers. The performance of 2nd model is improved. Either meidan cut-off or mean cut-off shows higher than 93% accuracy.

4. For cut-off values, by comparing 3 models' results, I prefer median value as cut-off since median won't be affected by sample size.

5. When it comes to the 3rd models, since 2nd model proves LSTM has better performance, I add a convolutional layer and dropout function on top of the LSTM layer. While even the accuracy on predicting the class0 was improved to 96% (for 2nd model it's 95%), accuracy of class 1 dropped to 87% compare to 93% in 2nd model. So 2nd will be the best among these three models.